affordance_primitives/include/affordance_primitives/screw_planning/screw_planning.hpp

```diff
@@ -36,22 +36,71 @@
 
 #include <affordance_primitives/screw_model/affordance_utils.hpp>
 #include <affordance_primitives/screw_model/screw_axis.hpp>
+#include <algorithm>
 #include <utility>
-
 namespace affordance_primitives
 {
-bool constraintFn(
-  const Eigen::Isometry3d & current_pose, const Eigen::Isometry3d & start_pose,
-  const ScrewAxis & screw_axis, double theta_max, double theta_guess,
-  Eigen::Ref<Eigen::VectorXd> out);
+/**
+* @brief Computes the set of screw angles that minimize the error between EE pose and screw path 
+*
+* @param tf_m_to_q Forward kinematics of the robot
+* @param tf_m_to_s Start pose
+* @param screw_axis_set Set of screw axes (S1,...,Sm) in frame M
+* @param phi_max Set of screw distances to move
+* @param phi_start Set of initial random guesses for screw angles
+*
+* @return Constraint function success, error, and corresponding screw angles
+*/
+std::tuple<bool, Eigen::VectorXd, Eigen::VectorXd> constraintFn(
+  const Eigen::Isometry3d & tf_m_to_q, const Eigen::Isometry3d & tf_m_to_s,
+  const std::vector<ScrewAxis> & screw_axis_set, const Eigen::VectorXd phi_max,
+  const Eigen::VectorXd phi_start);
+
+/**
+ * @brief Computes the partial derivative of error between EE pose and screw path
+ *
+ * @param tf_q_to_m Inverse of the forward kinematics matrix, tf_m_to_q
+ * @param tf_m_to_s Transform from base frame, M to screw path
+ * @param phi_current Current set of screw angles
+ * @param screw_axis_set Set of screw axes (S1,...,Sm) in frame M
+ *
+ * @return Vector of partial derivative of error with respect of screw angles
+ */
+Eigen::VectorXd errorDerivative(
+  const Eigen::Isometry3d & tf_q_to_m, const Eigen::Isometry3d & tf_m_to_s,
+  const Eigen::VectorXd & phi_current, const std::vector<ScrewAxis> & screw_axis_set);
 
-Eigen::VectorXd calcError(const Eigen::Isometry3d & tf_err);
+/**
+ * @brief Computes the 6x1 error vector for a given transformation matrix
+ *
+ * @param tf Transformation matrix
+ *
+ * @return 6x1 Error vector
+ */
+Eigen::VectorXd eta(const Eigen::Isometry3d & tf);
```

Comment:
How about error_eta? When I started with the 1D body-frame code, initially, I was totally lost because of the variable (and function) names. I suspect the following two reasons for that: 1) Although they were descriptive, descriptions can be subjective. 2) They were all different from the paper and that meant I couldn't effectively use the paper as reference. So when I wrote this, I decided to address these issues by adopting all names from the paper and keeping the general structure the same as the paper algorithms. This helped me a lot, and I think this helps build a consistent reference for you and me both. For the future readers and users of the code (besides us), we can make the Doxygen documentation very descriptive and add the algorithms and images to the README? Thoughts?

Reply:
I think I might go with `calculateEta()` or something, I do tend to like verbs in function names

Reply:
perfect!